### PR TITLE
feat: add mailgun service for sending mails

### PR DESCRIPTION
**Pull request by LakshayaT**
This PR adds and implements the mailgun service functionality for sending emails. Required code changes, updates and additions have been made.

Related changes:
- Add mailgun.js npm package
- Update .env.example to store mailgun related env variables Update notifications constant file to add MAILGUN type Add mailgun service files for sending mail
- Add mailgun queue file to store its config
- Add mailgun consumer for processing its queue
- Update notifications producer to add notifications to correct queue based on channelType Update notifications module to register queue, add mailgun providers

Sample request body:
```json
{
  "channelType": 2,
  "data": {
    "from": "sender@email.com",
    "to": "recepient1@email.com, recepient2@email.com",
    "subject": "Test subject",
    "text": "This is a test notification",
    "html": "<b>This is a test notification</b>"
  }
}
```

Sample response:
```json
{
  "notification": {
    "channelType": 2,
    "data": {
      "from": "sender@email.com",
      "to": "recepient1@email.com, recepient2@email.com",
      "subject": "Test subject",
      "text": "This is a test notification",
      "html": "<b>This is a test notification</b>"
    },
    "createdBy": "testuser",
    "updatedBy": "testuser",
    "result": null,
    "id": 21,
    "deliveryStatus": 1,
    "createdOn": "2023-09-04T12:31:21.000Z",
    "updatedOn": "2023-09-04T12:31:21.000Z",
    "status": 1
  }
}
```

### CODE REVIEW

File: src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  constructor(
+    @InjectRepository(Notification)
+    private readonly notificationRepository: Repository<Notification>,
+    private readonly mailgunService: MailgunService,
+    private readonly notificationsService: NotificationsService,
+  ) {}
+
+  @Process()
+  async processMailgunNotificationQueue(job: Job): Promise<void> {
```
Comment (xixas) [inline src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts:25]: `Job<number>`

File: src/jobs/producers/notifications/notifications.job.producer.ts [hunk 2]
```diff
@@ -3,18 +3,34 @@ import { InjectQueue } from '@nestjs/bull';
...
+    @InjectQueue(MAILGUN_QUEUE) private readonly mailgunQueue: Queue,
+  ) {}
   private readonly logger = new Logger(NotificationQueueProducer.name);
   async onModuleInit(): Promise<void> {
     this.smtpQueue.client.on('error', (error) => {
       this.logger.error('Redis connection error:');
       this.logger.error(JSON.stringify(error, ['message', 'stack', 2]));
     });
+    this.mailgunQueue.client.on('error', (error) => {
+      this.logger.error('Redis connection error:');
```
Comment (xixas) [inline src/jobs/producers/notifications/notifications.job.producer.ts:22]: No need to add an error again. Checking for Redis in one queue is enough.

File: src/services/email/mailgun/mailgun.service.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    this.mailgunClient = this.mailgun.client({
+      username: 'api',
+      key: this.configService.getOrThrow<string>('MAILGUN_API_KEY'),
+      host: this.configService.getOrThrow<string>('MAILGUN_HOST'),
+    } as MailgunClientOptions);
+  }
+
+  sendEmail(mailgunNotificationData: MailgunMessageData): Promise<MessagesSendResult> {
+    return this.mailgunClient.messages.create(
+      this.configService.getOrThrow<string>('MAILGUN_DOMAIN'),
```
Comment (xixas) [inline src/services/email/mailgun/mailgun.service.ts:22]: No need to call ` this.configService.getOrThrow<string>('MAILGUN_DOMAIN'),` inside function. Keep it as a private string variable like `mailgunDomain` and assign value in constructor. Can be reused if needed later by other service methods.

**Review by xixas — CHANGES_REQUESTED** [review]

File: src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts (hunk 1 above)
Comment (LakshayaT) [reply src/jobs/consumers/notifications/mailgun-notifications.job.consumer.ts]: This has been changed in the latest commit

File: src/jobs/producers/notifications/notifications.job.producer.ts (hunk 2 above)
Comment (LakshayaT) [reply src/jobs/producers/notifications/notifications.job.producer.ts]: Removed the error logging for redis from mailgun queue in latest commit

File: src/services/email/mailgun/mailgun.service.ts (hunk 3 above)
Comment (LakshayaT) [reply src/services/email/mailgun/mailgun.service.ts]: Created new private string variable mailgunDomain and store value from env in it in latest commit.

**Review by xixas — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -34,6 +34,7 @@
     "bull": "^4.11.3",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.0",
+    "mailgun.js": "^9.2.1",
```
Comment (osm-vishnukyatannawar) [inline package.json:37]: Is this official library?

**Review by osm-vishnukyatannawar — APPROVED** [review]

File: package.json (hunk 4 above)
Comment (osm-vishnukyatannawar) [reply package.json:37]: cross verified it.

**Pull request merged**